File: java/1608. Special Array With X Elements Greater Than or Equal X.java
// Time Complexity : O(n logn), where n is the length of nums. Sorting takes O(n logn) time. After sorting, the nums array is traversed, with constant time operations. So traversing the nums array takes O(n) time. Overall time complexity is O(n logn) time.
// Space Complexity : O(n logn), sorting takes O(n logn) time, and there are no other data structure used except constants. 

class Solution {
    public int specialArray(int[] nums) {
        int currMax = -1;
        Arrays.sort(nums);
        int l = nums.length;
        for(int i = 0; i < nums.length; i++){
            int num = nums[i];
            if(nums[i] >= l && currMax < l){
                return l;
            }
            l--;
            currMax = num;
        }
        return -1;
    }
}

File: python/1608. Special Array With X Elements Greater Than or Equal X.py
# Time Complexity : O(n logn), where n is the length of nums. Sorting takes O(n logn) time. After sorting, the nums array is traversed, with constant time operations. So traversing the nums array takes O(n) time. Overall time complexity is O(n logn) time.
#  Space Complexity : O(n logn), sorting takes O(n logn) time, and there are no other data structure used except constants. 

class Solution:
    def specialArray(self, nums: List[int]) -> int:
        nums.sort()
        lastMax = -1
        l = len(nums)
        for i in nums:
            if i >= l and l > lastMax:
                return l
            else:
                l -= 1
            lastMax = i
        return -1